playbooks/migrating-postgres-with-rubyrep.md

```diff
@@ -0,0 +1,375 @@
+  template:
+    spec:
+      affinity:
+        nodeAffinity:
+          preferredDuringSchedulingIgnoredDuringExecution:
+          - preference:
+              matchExpressions:
+              - key: tier
+                operator: In
+                values:
+                - background
+            weight: 1
+      containers:
+      - args:
+        - sh
+        - ./import-db.sh
+        - { environment }
+        - "--clean --no-owner --schema=public -j 5 -v"
+        env:
+        - name: APP_NAME
+          value: { myapp }
+        - name: DATABASE_URL
+          value: { new database url }
+        - name: AWS_ACCESS_KEY_ID
+          value: { app aws access key id }
+        - name: AWS_SECRET_ACCESS_KEY
+          value: { app secret access key }
+        - name: SWALLOW_ERRORS_ON_RESTORE
+          value: "1"
+        image: artsy/pg-data-sync
+        imagePullPolicy: Always
+        name: { myapp }-data-import
+      dnsPolicy: ClusterFirst
+      restartPolicy: Never
+      terminationGracePeriodSeconds: 30
+
+```
+
+Apply with `hokusai [staging|production] create --filename ./hokusai/import-to-new-production.yml`
+
+Once complete, delete with `hokusai [staging|production] delete --filename ./hokusai/import-to-new-production.yml`
+
+Note: if you are unable to backup the old database and restore to the new via the `pg-data-sync` image (for instance in the case of differing Postgres major versions) you can load the app schema into the new database instead, and let RubyRep perform the migration.
+
+Start a shell in the application with the new database connection and load the app schema with:
+```
+hokusai production run --env "DATABASE_URL=postgres://{new database credentials}" "bundle exec rake db:schema:load"
+```
+
+However, be aware the RubyRep is known to deadlock in some cases when migrating large tables ( over 1Gb in size ) so skipping the initial dump and restore should only be used as a fallback!
+
+3) Log into the new database and note all foreign key constraints.  You can get foreign key constraints for each table with `\d+ table` or to see forign key constraints for all tables in the `public` schema, run:
+```
+SELECT conrelid::regclass AS table_from
+     , conname
+     , pg_get_constraintdef(oid)
+FROM   pg_constraint
+WHERE  contype IN ('f')
+AND    connamespace = 'public'::regnamespace  -- your schema here
+ORDER  BY conrelid::regclass::text, contype DESC;
+```
+
+Example output:
+```
+      table_from     |       conname       |                    pg_get_constraintdef
+--------------------+---------------------+-------------------------------------------------------------
+foo                  | fk_rails_fbc9d01ca0 | FOREIGN KEY (bar_id) REFERENCES bar(id)
+bar                  | fk_rails_80eb82ccbf | FOREIGN KEY (foo_id) REFERENCES foo(id) ON DELETE CASCADE
+```
+
+Save the constraints to later re-enable.
+
+4) Disable all foreign key constraints, for example for tables `foo` and `bar` you would run:
```

Comment:
@eessex @joeyAghion @mzikherman thinking that this could be automated via a fancy PSQL script or (since my PSQL isn't so great) perhaps baked into RubyRep as a command, i.e. `rubyrep disable_fk_constraints`